src/state_machine.zig

```diff
@@ -505,6 +555,97 @@ pub fn StateMachineType(
+                    // so the user can use the last seen timestamp to get the next block of results:
+                    .ascending => TimestampRange.gte(filter.timestamp + 1),
+                    .descending => TimestampRange.lte(filter.timestamp - 1),
+                };
+
+                // This query may have 2 conditions:
+                // `WHERE debit_account_id = $account_id OR credit_account_id = $account_id`.
+                var scan_conditions: stdx.BoundedArray(*TransfersGroove.ScanGroove.Scan, 2) = .{};
+
+                // If no flag was set, we default to include both credits and debits.
+                const include_all = filter.flags.debits == false and filter.flags.credits == false;
+
+                // Adding the condition for `debit_account_id = $account_id`.
+                if (filter.flags.debits or include_all) {
+                    scan_conditions.append_assume_capacity(scan_groove.scan_prefix(
+                        .debit_account_id,
+                        self.forest.scan_buffer_pool.acquire_assume_capacity(),
+                        snapshot_latest,
+                        filter.account_id,
+                        timestamp_range,
+                        direction,
+                    ));
+                }
+
+                // Adding the condition for `credit_account_id = $account_id`.
+                if (filter.flags.credits or include_all) {
+                    scan_conditions.append_assume_capacity(scan_groove.scan_prefix(
+                        .credit_account_id,
+                        self.forest.scan_buffer_pool.acquire_assume_capacity(),
+                        snapshot_latest,
+                        filter.account_id,
+                        timestamp_range,
+                        direction,
+                    ));
+                }
+
+                break :scan switch (scan_conditions.count()) {
+                    1 => scan_conditions.get(0),
+                    // Creating an union `OR` with the conditions.
+                    2 => scan_groove.merge_union(
+                        scan_conditions.const_slice(),
+                        direction,
+                    ),
+                    else => unreachable,
+                };
+            };
+
+            // Creating a fetcher to materialize the Transfers found by the scan:
+            const fetcher = scan_groove.fetch(scan, snapshot_latest);
+            const context = self.scan_context.get(.transfers, &scan_account_transfers_callback);
+            const buffer = std.mem.bytesAsSlice(Transfer, self.scan_buffer);
+
+            fetcher.read(
+                context,
+                // Limiting the buffer size according to the query limit,
+                // No limit if it is ZERO or exceeds the buffer size.
+                if (filter.limit == 0 or filter.limit > buffer.len)
+                    buffer[0..]
+                else
+                    buffer[0..filter.limit],
+            );
```

Comment:
> rather than just putting results directly into the reply message body

I implemented it the same way `lookup_*` works, starting the scan during prefetch and copying the results from the buffer into the reply message body on commit.
For lookups the groove's `ObjectsCache` acts like this buffer.

Are you suggesting that we can access the reply message before `commit`?

Reply:
> Are you suggesting that we can access the reply message before commit?

No, this comment wasn't a suggestion at all, I was just checking my understanding of why/how it works the way it does.
I don't think we can access the reply message prior to `commit`.

Reply:
> I was just checking my understanding of why/how it works the way it does.

I don't think I got the other way it could be done 😉

> it allows us to start the scan during prefetch, rather than commit?

It never occurred to me the possibility of starting the scan during the commit phase.
Is it even possible to run async code inside `commit()`?